cpp/src/Ice/EndpointFactoryManager.cpp

```diff
@@ -89,8 +97,12 @@ IceInternal::EndpointFactoryManager::create(const string& str, bool oaEndpoint)
 
     EndpointFactoryPtr factory;
     {
-        lock_guard lock(_mutex); // TODO: Necessary?
-
+        // Hold the mutex while looking up the factory, destroy can be called concurrently.
```

Comment:
I don't see how this PR solves the race condition.

When you call the factory later on in this function, it's outside the lock, and the factory code itself is not thread-safe. In particular, we could be destroying the factory (setting its _instance to nullptr - see TcpEndpointFactory) and using this factory concurrently.

In my view, the only proper way to solve this issue is to use weak_ptr.

Reply:
Updated the code to use a weak_ptr.